redash/handlers/query_results.py

```diff
@@ -318,11 +318,16 @@ def get(self, query_id=None, query_result_id=None, filetype="json"):
         if query_id is not None:
             query = get_object_or_404(models.Query.get_by_id_and_org, query_id, self.current_org)
 
-            if query_result is None and query is not None and query.latest_query_data_id is not None:
-                query_result = get_object_or_404(
-                    models.QueryResult.get_by_id_and_org,
-                    query.latest_query_data_id,
-                    self.current_org,
+            if query_result is None and query is not None:
+                # query.latest_query_data_id isn't db_role-specific, so
+                # ignore it and fetch the latest results for the current
+                # user's role.
+                query_result = models.QueryResult.get_latest(
+                    data_source=query.data_source,
+                    query=query.query_hash,
+                    max_age=-1,
+                    is_hash=True,
+                    db_role=getattr(self.current_user, "db_role", None),
                 )
```

Comment:
You still need the `get_object_or_404` wrapper, but I think it should be straightforward. Something like:
```suggestion
                query_result = get_object_or_404(
                    models.QueryResult.get_latest,
                    data_source=query.data_source,
                    query=query.query_hash,
                    max_age=-1,
                    is_hash=True,
                    db_role=getattr(self.current_user, "db_role", None),
                )
```

I'm a little surprised the existing tests don't include a 404 case. Either way it's probably worth adding one for the user / admin case (user with only an admin result, admin with only a user result).

Reply:
Hmm I actually like the user experience of this better without the `get_object_or_404()`. :eyes: 

With the change in this PR, we get an empty query result which aborts [later in the same method](https://github.com/stacklet/redash/blob/a19b166099ea2c7399e7d4a01a23525ed3e603ce/redash/handlers/query_results.py#L382-L383) with a 404 status code and the message:

> No cached result found for this query.

If we add the `get_object_or_404()` wrapper, the caller still gets a 404 response but with a more generic message:

> The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

I _think_ the more specific response message is useful for what we're doing here, though it's not a super strong opinion :thinking: . Does this context affect your thoughts one way or the other?

Reply:
Meh nevermind. I'm outnumbered here by you and Cursor :sweat_smile: , and didn't have a particularly strong aversion to the wrapper anyway. Just added it.

Reply:
Ah, I missed that it was handled in that other bit and the wrapper does make that handling a bit redundant, but since it was already like that, I guess it's better to keep the existing behavior? 🤷